server/models/services/TestPlanReportService.js

```diff
@@ -0,0 +1,491 @@
+ */
+const testResultAssociation = testResultAttributes => ({
+    association: 'testResults',
+    attributes: testResultAttributes
+});
+
+/**
+ * NB. You can pass any of the attribute arrays as '[]' to exclude that related association
+ * @param {number} id - unique id of the TestPlanReport model being queried
+ * @param {string[]} testPlanReportAttributes - TestPlanReport attributes to be returned in the result
+ * @param {string[]} testPlanRunAttributes - TestPlanRun attributes to be returned in the result
+ * @param {string[]} testPlanAttributes - TestPlan attributes to be returned in the result
+ * @param {string[]} testPlanTargetAttributes - TestPlanTarget attributes to be returned in the result
+ * @param {string[]} userAttributes - User attributes to be returned in the result
+ * @param {string[]} testResultAttributes - TestResult attributes to be returned in the result
+ * @returns {Promise<*>}
+ */
+const getTestPlanReportById = async (
+    id,
+    testPlanReportAttributes = TEST_PLAN_REPORT_ATTRIBUTES,
+    testPlanRunAttributes = TEST_PLAN_RUN_ATTRIBUTES,
+    testPlanAttributes = TEST_PLAN_ATTRIBUTES,
+    testPlanTargetAttributes = TEST_PLAN_TARGET_ATTRIBUTES,
+    userAttributes = USER_ATTRIBUTES,
+    testResultAttributes = TEST_RESULT_ATTRIBUTES
+) => {
+    // check to see if 'canonicalRun' has been requested; custom attribute handling required
+    const canonicalRunKeyFound = testPlanReportAttributes.some(
+        key => key === 'canonicalRun'
+    );
+
+    // required to be used to determine whether or not to actually return 'canonicalRun'; TestPlan.publishStatus MUST be 'final' for 'canonicalRun' to be returned
+    const testPlanReportPublishStatusFound = testPlanReportAttributes.some(
+        key => key === 'publishStatus'
+    );
+    const testPlanParsedFound = testPlanAttributes.some(
+        key => key === 'parsed'
+    );
+    const testPlanRunAttributesFound = testPlanAttributes.length;
+
+    if (canonicalRunKeyFound) {
+        testPlanReportAttributes = [
+            ...testPlanReportAttributes,
+            'publishStatus'
+        ];
+        testPlanAttributes = [...testPlanAttributes, 'parsed'];
+        testPlanRunAttributes = [...testPlanRunAttributes, 'tester'];
+    }
+
+    // filter out 'canonicalRun' so it doesn't pollute attributes being passed to sequelize structure
+    testPlanReportAttributes = testPlanReportAttributes.filter(
+        key => key !== 'canonicalRun'
+    );
+
+    let result = await ModelService.getById(
+        TestPlanReport,
+        id,
+        testPlanReportAttributes,
+        [
+            testPlanRunAssociation(
+                testPlanRunAttributes,
+                userAttributes,
+                testResultAttributes
+            ),
+            testPlanAssociation(testPlanAttributes),
+            testPlanTargetAssociation(testPlanTargetAttributes)
+        ]
+    );
+
+    if (canonicalRunKeyFound) {
```

Comment:
Would definitely like some special focus on the handling of `canonicalRun` @alflennik and what it could mean for the format coming from the `TestResult.data`. Haven't fully locked down how it's handled, whether at this layer (which is ok too) or as some virtual property of the TestPlanReport model so was considering to remove support for it until we've confirmed the `TestResult.data`. Thoughts?

Reply:
@howard-e It can be any of the runs, maybe the first one, or a random one. When the report is finalized that means all conflicts between the results have been resolved, so they're all identical.

Reply:
There's no need for an additional key.

Reply:
That means this can just be easily handled in the resolver then? To just move up any object found in the linked `testPlanRuns` array as the `canonicalRun`